--- a/src/orca-server-mahimahi.cc	2022-06-16 18:40:20.540490110 +0000
+++ b/src/orca-server-mahimahi.cc	2022-06-16 22:35:25.179604318 +0000
@@ -126,7 +126,7 @@
 
         int reuse = 1;
         if (setsockopt(sock[i], SOL_SOCKET, SO_REUSEADDR, (const char*)&reuse, sizeof(reuse)) < 0)
-            perror("setsockopt(SO_REUSEADDR) failed");
+            perror("setsockopt(SO_REUSEADDR) failed \n");
         //Bind socket on IP:Port
         if(bind(sock[i],(struct sockaddr *)&server_addr[i],sizeof(struct sockaddr))<0)
         {
@@ -145,25 +145,25 @@
     }
 
     char container_cmd[500];
-    sprintf(container_cmd,"sudo -u `whoami` %s/client $MAHIMAHI_BASE 1 %d",path,client_port);
+    // sprintf(container_cmd,"sudo -u `whoami` %s/client $MAHIMAHI_BASE 1 %d",path,client_port);
     char cmd[1000];
     char final_cmd[1000];
 
-    if (first_time==4 || first_time==2)
-        sprintf(cmd, "sudo -u `whoami`   mm-delay %d mm-link %s/../traces/%s %s/../traces/%s --downlink-log=%s/log/down-%s --uplink-queue=droptail --uplink-queue-args=\"packets=%d\" --downlink-queue=droptail --downlink-queue-args=\"packets=%d\" -- sh -c \'%s\' &",delay_ms,path,uplink,path,downlink,path,log_file,qsize,qsize,container_cmd);
-    else
-        sprintf(cmd, "sudo -u `whoami`  mm-delay %d mm-link %s/../traces/%s %s/../traces/%s --uplink-queue=droptail --uplink-queue-args=\"packets=%d\" --downlink-queue=droptail --downlink-queue-args=\"packets=%d\" -- sh -c \'%s\' &",delay_ms,path,uplink,path,downlink,qsize,qsize,container_cmd);
+    // if (first_time==4 || first_time==2)
+    //     sprintf(cmd, "sudo -u `whoami`   mm-delay %d mm-link %s/../traces/%s %s/../traces/%s --downlink-log=%s/log/down-%s --uplink-queue=droptail --uplink-queue-args=\"packets=%d\" --downlink-queue=droptail --downlink-queue-args=\"packets=%d\" -- sh -c \'%s\' &",delay_ms,path,uplink,path,downlink,path,log_file,qsize,qsize,container_cmd);
+    // else
+    //     sprintf(cmd, "sudo -u `whoami`  mm-delay %d mm-link %s/../traces/%s %s/../traces/%s --uplink-queue=droptail --uplink-queue-args=\"packets=%d\" --downlink-queue=droptail --downlink-queue-args=\"packets=%d\" -- sh -c \'%s\' &",delay_ms,path,uplink,path,downlink,qsize,qsize,container_cmd);
     
-    sprintf(final_cmd,"%s",cmd);
+    // sprintf(final_cmd,"%s",cmd);
 
-    DBGPRINT(DBGSERVER,0,"%s\n",final_cmd);
+    // DBGPRINT(DBGSERVER,0,"%s\n",final_cmd);
     info->trace=trace;
     info->num_lines=num_lines;
     /**
      *Setup Shared Memory
      */ 
-    key=(key_t) (actor_id*10000+rand()%10000+1);
-    key_rl=(key_t) (actor_id*10000+rand()%10000+1);
+    key=(key_t) (actor_id*10000+rand()%10000+1 + (client_port*10000));
+    key_rl=(key_t) (actor_id*10000+rand()%10000+1 + ((client_port*10000)));
     // Setup shared memory, 11 is the size
     if ((shmid = shmget(key, shmem_size, IPC_CREAT | 0666)) < 0)
     {
@@ -189,16 +189,16 @@
         return;
     } 
     if (first_time==1){
-        sprintf(cmd,"/home/`whoami`/venv/bin/python %s/d5.py --tb_interval=1 --base_path=%s --task=%d --job_name=actor --train_dir=%s --mem_r=%d --mem_w=%d &",path,path,actor_id,path,(int)key,(int)key_rl);
+        sprintf(cmd,"/home/`whoami`/venv/bin/python -W ignore -W ignore::DeprecationWarning %s/d5.py --tb_interval=1 --base_path=%s --task=%d --job_name=actor --train_dir=%s --mem_r=%d --mem_w=%d &",path,path,actor_id,path,(int)key,(int)key_rl);
         DBGPRINT(0,0,"Starting RL Module (Without load) ...\n%s",cmd);
     }
     else if (first_time==2 || first_time==4){
-        sprintf(cmd,"/home/`whoami`/venv/bin/python %s/d5.py --tb_interval=1 --base_path=%s --load --eval --task=%d --job_name=actor --train_dir=%s  --mem_r=%d --mem_w=%d &",path,path,actor_id,path,(int)key,(int)key_rl);
+        sprintf(cmd,"/home/`whoami`/venv/bin/python -W ignore -W ignore::DeprecationWarning %s/d5.py --tb_interval=1 --base_path=%s --load --eval --task=%d --job_name=actor --train_dir=%s  --mem_r=%d --mem_w=%d &",path,path,actor_id,path,(int)key,(int)key_rl);
         DBGPRINT(0,0,"Starting RL Module (No learning) ...\n%s",cmd);
     }
     else
     {
-        sprintf(cmd,"/home/`whoami`/venv/bin/python %s/d5.py --load --tb_interval=1 --base_path=%s --task=%d --job_name=actor --train_dir=%s  --mem_r=%d --mem_w=%d &",path,path,actor_id,path,(int)key,(int)key_rl);
+        sprintf(cmd,"/home/`whoami`/venv/bin/python -W ignore -W ignore::DeprecationWarning %s/d5.py --load --tb_interval=1 --base_path=%s --task=%d --job_name=actor --train_dir=%s  --mem_r=%d --mem_w=%d &",path,path,actor_id,path,(int)key,(int)key_rl);
         DBGPRINT(0,0,"Starting RL Module (With load) ...\n%s",cmd);
     }
  
@@ -240,7 +240,7 @@
     DBGPRINT(0,0,"RL Module is Ready. Let's Start ...\n\n");    
     usleep(actor_id*10000+10000);
     //Now its time to start the server-client app and tune C2TCP socket.
-    system(final_cmd);
+    // system(final_cmd);
         
     //Start listen
     int maxfdp=-1;
@@ -532,7 +532,7 @@
            else{
                 if (error2_cnt==50)
                 {
-                    DBGPRINT(0,0,"got null values: (downlink:%s delay:%d qs:%d) Actor: %d iteration:%d\n",downlink,delay_ms,qsize,actor_id,step_it);
+                    // DBGPRINT(0,0,"got null values: (downlink:%s delay:%d qs:%d) Actor: %d iteration:%d\n",downlink,delay_ms,qsize,actor_id,step_it);
                     //FIXME:
                     //A Hack for now! Let's send a new state to get new action in case we have missed previous action. Why it happens?!
                     if((1+tmp_step)==(step_it))
@@ -674,10 +674,10 @@
 		{
 			DBGMARK(DBGSERVER,5,"++++++\n");
 			len-=send(sock_local,write_message,strlen(write_message),0);
-		    usleep(50);         
+		    // usleep(50);         
             DBGMARK(DBGSERVER,5,"------\n");
 		}
-        usleep(100);
+        // usleep(100);
 	}
 	flow->flowinfo.rem_size=0;
     done=true;
